### ISSUE TITLE
[HF transformers] _ignore_causal_mask_sdpa blocks SDPA causal shortcut under torch.compile -> non-bitwise output vs eager

### ISSUE THREAD

**Issue by penguinwu** [penguinwu/oss-model-graph-break-corpus#129]
## Summary

`transformers.masking_utils._ignore_causal_mask_sdpa()` (and its sibling `_ignore_bidirectional_mask_sdpa()`) take a **different code path under `torch.compile` than under eager execution**, which breaks bitwise equivalence between `model(x)` and `torch.compile(model, backend="eager", fullgraph=False)(x)`.

- **Eager**: when no token is masked and `query_length == 1` or `kv_length == query_length`, the helper returns `True` → an explicit attention mask is *not* materialized → SDPA is called with `is_causal=True, attn_mask=None` and can use GQA + the flash/causal kernel.
- **Compile**: `is_tracing(padding_mask)` is `True` under Dynamo, so the helper returns `False` → an explicit boolean mask is materialized → SDPA is called with `is_causal=False, attn_mask=<bool tensor>`, K/V are manually repeated, and GQA is disabled.

PyTorch's SDPA dispatcher selects a **different kernel** for those two argument sets, and the two kernels produce numerically different output (order `1e-6` to `1e-5` per layer). Since `backend="eager"` is meant to be a no-op backend, any divergence from eager output is a correctness gap.

In a sweep of ~190 HF model classes under `torch.compile(backend="eager", fullgraph=False)`, this single mechanism accounts for **~92% of all non-bitwise-equal (model, mode) pairs**.

## Environment

- Sweep / reproduction ran on transformers 5.6.2, but the buggy code path is **unchanged in every release through 5.9.0 (the current latest, released 2026-05-20)** — verified directly against the v5.9.0 source: the guard is at `masking_utils.py` L293 (`_ignore_causal_mask_sdpa`) and L354 (sibling `_ignore_bidirectional_mask_sdpa`). The fix applies cleanly to current `main`.
- torch 2.13 nightly (cu126); reproduces on stable torch ≥ 2.3
- CUDA GPU

## Reproduction

```python
import torch, torch._dynamo
from transformers import BartModel, BartConfig

torch.manual_seed(0)
model = BartModel(BartConfig()).cuda().eval()
x = torch.randint(0, 1000, (1, 16)).cuda()

torch._dynamo.reset(); torch.manual_seed(0)
with torch.no_grad():
    out_eager = model(x).last_hidden_state

torch._dynamo.reset(); torch.manual_seed(0)
compiled = torch.compile(model, backend="eager", fullgraph=False)
with torch.no_grad():
    out_compile = compiled(x).last_hidden_state

print("bitwise equal:", torch.equal(out_eager, out_compile))           # False
print("max abs diff :", (out_eager - out_compile).abs().max().item())  # ~1e-6
```

Confirming the mechanism — force the SDPA MATH backend on **both** paths (so the same kernel runs on both) and the output becomes bitwise-identical:

```python
from torch.nn.attention import sdpa_kernel, SDPBackend
with sdpa_kernel([SDPBackend.MATH]):
    # re-run both the eager and the compiled forward pass inside this context
    ...
# torch.equal(out_eager, out_compile) -> True, max abs diff 0.0
```

## Root cause

`_ignore_causal_mask_sdpa()` in `transformers/masking_utils.py`:

```python
if (
    not is_tracing(padding_mask)        # <-- blocks the shortcut under torch.compile
    and (query_length == 1 or kv_length == query_length)
    and (local_attention_size is None or kv_length < local_attention_size)
    and (padding_mask is None or padding_mask.all())
):
    return True
return False
```

`is_tracing()` (in `transformers/utils/import_utils.py`) returns `True` for **both** `torch.compile` (Dynamo) and `torch.export`:

```python
_is_tracing = is_torchdynamo_compiling() or is_jit_tracing() or is_cuda_stream_capturing()
```

The guard exists to protect **`torch.export`**: export hard-codes `is_causal` into the exported graph, which is wrong for `query_length > 1` (see pytorch/pytorch#108108). But the *other* conditions in the same `if` (`query_length == 1 or kv_length == query_length`) already restrict the shortcut to exactly the cases where `is_causal=True` is correct, and **`torch.compile` does not hard-code `is_causal`** — so blocking the shortcut under plain `torch.compile` is unnecessary, and it is precisely what breaks eager-equivalence.

The same `not is_tracing(...)` guard is present in the sibling `_ignore_bidirectional_mask_sdpa()` (same file) and should be fixed consistently. The XPU helpers `_can_skip_causal_mask_xpu` / `_can_skip_bidirectional_mask_xpu` have an analogous `if is_tracing(...): return False` early-out that is worth reviewing for the same reason.

## Proposed fix

HF already ships an export-only predicate, `is_torchdynamo_exporting()` (wraps `torch.compiler.is_exporting()`), which is distinct from `is_torchdynamo_compiling()` (wraps `torch.compiler.is_compiling()`, which is `True` for both compile **and** export). Allow the shortcut under compile while keeping it blocked under export and the other tracing modes:

```python
from .utils.import_utils import is_torchdynamo_compiling, is_torchdynamo_exporting

# torch.export hard-codes is_causal into the graph (wrong for query_length > 1,
# pytorch/pytorch#108108). torch.compile has no such constraint and must match eager.
is_compile_not_export = is_torchdynamo_compiling() and not is_torchdynamo_exporting()
if (
    (not is_tracing(padding_mask) or is_compile_not_export)
    and (query_length == 1 or kv_length == query_length)
    and (local_attention_size is None or kv_length < local_attention_size)
    and (padding_mask is None or padding_mask.all())
):
    return True
return False
```

**Simpler alternative** (if maintainers are comfortable also enabling the shortcut under `torch.jit` / `torch.fx`, which the inner conditions already make numerically safe): replace `not is_tracing(padding_mask)` with `not is_torchdynamo_exporting()`.

Apply the same change to `_ignore_bidirectional_mask_sdpa()`.

## Validation

1. Run the reproduction above; assert `torch.equal(out_eager, out_compile)` is `True` after the patch.
2. Regression sweep: the fix was validated (via the "SDPA MATH on both paths" surrogate, which forces the same kernel on eager and compile) to produce bitwise-equal output across 30+ model classes spanning encoder-decoder seq2seq (Bart, MBart, Marian, Pegasus, Blenderbot, M2M100, PLBart), MoE / hybrid-Mamba causal-LMs (Jamba, NemotronH, Dbrx, Qwen3Next, OlmoHybrid, FalconH1), multimodal VLMs (Gemma4, Lfm2Vl, several OCR-VLMs), and audio encoders (Wav2Vec2, Hubert, UniSpeech, Data2VecAudio) in eval mode.
3. Confirm `torch.export` behavior is **unchanged**: export a model with `query_length > 1` and verify the explicit mask is still materialized (the export path must keep returning `False`).

## Impact

Resolves ~111 of 121 non-bitwise-equal rows (~92%) in the HF-model sweep under `backend="eager"`. The remaining gap is a separate train-mode LayerDrop/Dropout RNG issue (filed as a companion issue).
